### ISSUE TITLE
Escape text input

### ISSUE THREAD

**Issue by yoshuawuyts** [yoshuawuyts/html#12]
Right now we're not doing any escaping, but we probably should.

Maybe instead of a single `text` API, we should expose two?

- `text`: takes strings and escapes them 
- `raw_text`: takes strings and doesn't escape them 

Thoughts?

**Comment by yoshuawuyts** [comment]
Oh actually maybe we should make all of our string APIs take a `text` element which auto-escapes on input. This can hold a Cow internally, but should be nicer to use. And it allows text to be escaped just once and reused.

**Comment by yoshuawuyts** [comment]
Yeah, if we ever want to style text elements (https://github.com/yoshuawuyts/html/issues/14) we will want text to be individually addressable.

This probably means the `.text("foo")` constructor should take a closure as a second argument?


```rust
li.text("hello", |text| text.color(|c| c.red()));
```

**Comment by yoshuawuyts** [comment]
Bikeshed names: `styled_text`

**Comment by henkkuli** [comment]
Here's a new idea that allows the API to be safe and escaping by default, but also expressive and expandable if needed: Add a new trait `IntoHtmlText` that takes care of properly escaping the text:
```rust
pub trait IntoHtmlText {
    fn into_html_text(self) -> Cow<'static, str>;
}
```
This crate would provide default implementations for `str`-like elements:
```rust
impl IntoHtmlText for String {
    fn into_html_text(self) -> Cow<'static, str> {
        escape_html(self)
    }
}

impl IntoHtmlText for &'static str {
    fn into_html_text(self) -> Cow<'static, str> {
        escape_html(self)
    }
}

// and others like Cow ...
```
To allow using raw HTML, the crate would also provide a simple wrapper type `RawHtml`, possibly with a more scarry-sounding name:
```rust
/// Raw HTML that has been pre-formatted.
///
/// Note that this crate does not and cannot check whether the string you give
/// has been correctly formatted. It is your responsibility to ensure that the
/// given string is valid HTML in the given context.
pub struct RawHtml(pub String);

impl IntoHtmlText for RawHtml {
    fn into_html_text(self) -> Cow<'static, str> {
        self.0.into()
    }
}
```
Finally, all elements taking text would just take `impl IntoHtmlText` as parameter:
```rust
pub fn text(&mut self, t: impl IntoHtmlText) -> &mut Self {
    self.element.children_mut().push(t.into_html_text());
    self
}
```

Now the usage of this new API would look something like this:
```rust
p.text("You can style HTML elements with tags: <span style='color: red'>this text is red</span>")
 .text(RawHtml("The above example would render as <span style='color: red'>this text is red</span>"));
```
This can even be expanded later on to allow stylized text if #14 gets implemented:
```rust
p.text(TextFormatter::builder().color("red").text("red text").underline("and now with underline").build());
```

**Comment by henkkuli** [comment]
And now that I think of it, `IntoHtmlText` should implement an API based on `fmt::Formatter` just like `Debug`, `Display` and `Render`. Though it should be different from `Render` in the sense that external crates can also implement it, say for example a HTML parser crate.
